In [14]:
type(False)
type(2) # by default, python treats whole numbers as integers
type(2.0)  # to force it not to be an integer, add a .0
type("Hello, programmer!")

str

In [15]:
message = "So long and thanks for all the fish"
year = 2025
the_answer = 42
earth_demolished = False

In [19]:
message + ", sang the dolphins."

year + the_answer

not earth_demolished

True

In [22]:
#1st_thing = "No starting with numbers"

#first~thing = "No other symbols"

first.thing = "Periods have a particular meaning!"

NameError: name 'first' is not defined

In [34]:
type(2 + 3.14159) # add integer 2 and pi
#type(2 + True) # add integer 2 and TRUE
#type(True + False) # add TRUE and FALSE
#type(3.777 + True)


float

In [40]:
x = 3
y = "3.14159"
yy = float(y)
#x + y
x + yy
#x + float(y)

6.14159

In [42]:
14 // 3
14 % 3

4

In [57]:
import pandas as pd

dat = pd.read_csv("https://gist.githubusercontent.com/slopp/ce3b90b9168f2f921784de84fa445651/raw/4ecf3041f0ed4913e7c230758733948bc561f434/penguins.csv")

dat

# dat.info()


rowid    species     island  bill_length_mm  ...  flipper_length_mm  body_mass_g     sex  year
0        1     Adelie  Torgersen            39.1  ...              181.0       3750.0    male  2007
1        2     Adelie  Torgersen            39.5  ...              186.0       3800.0  female  2007
2        3     Adelie  Torgersen            40.3  ...              195.0       3250.0  female  2007
3        4     Adelie  Torgersen             NaN  ...                NaN          NaN     NaN  2007
4        5     Adelie  Torgersen            36.7  ...              193.0       3450.0  female  2007
..     ...        ...        ...             ...  ...                ...          ...     ...   ...
339    340  Chinstrap      Dream            55.8  ...              207.0       4000.0    male  2009
340    341  Chinstrap      Dream            43.5  ...              202.0       3400.0  female  2009
341    342  Chinstrap      Dream            49.6  ...              193.0       3775.0    male  2009
342    343  Chinstrap      Dream            50.8  ...              210.0       4100.0    male  2009
343    344  Chinstrap      Dream            50.2  ...              198.0       3775.0  female  2009

[344 rows x 9 columns]

In [60]:
## using methods
dat.body_mass_g.mean()

## editing elements
dat.body_mass_g[0] = 10000000
dat.body_mass_g

## boolean masking
big_penguins = dat.body_mass_g > 6000
dat
dat.loc[big_penguins]

<positron-console-cell-60>:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy


rowid species     island  bill_length_mm  ...  flipper_length_mm  body_mass_g   sex  year
0        1  Adelie  Torgersen            39.1  ...              181.0   10000000.0  male  2007
169    170  Gentoo     Biscoe            49.2  ...              221.0       6300.0  male  2007
185    186  Gentoo     Biscoe            59.6  ...              230.0       6050.0  male  2007

[3 rows x 9 columns]

## Overview

In this PA, you will will be creating and manipulating basic objects to uncover a top secret message.

### Hint 1

As you work through this PA, you will encounter some code that does not work as you want it to.  Don't despair!  Errors (when R is unable to run your code) and bugs (when the code doesn't do what you hoped) are a natural part of coding.  Even the best of the best deal with these problems regularly - learning to track down the issue is a skill that you can learn and practice.

Google is your friend!  If an error is confusing, copy it into Google and see what other people are saying.  If you don't know how to do something, search for it.

### Hint 2

Just because there is no error message doesn't mean everything went smoothly.  You should be regularly printing out your objects, to see they contain the information and format that you expect.

For example, if you have an object named `my_thing`, you might first want to simply look at the object by typing its name.

However, you might also want to make sure it is the right **object type** and/or the right **object structure**.


In [2]:
type(my_thing)
len(my_thing)

NameError: name 'my_thing' is not defined

## Part One: Setup

Each of the following code chunks will cause an error and/or do the desired task incorrectly.  Find the mistake, and correct it to complete the intended action.




1. Create lists containing the upper case letters, lower case letters, and some punctuation marks.


In [61]:
lower_case = ["a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "k", "l", "m", "n", "o", "p", "q", "r", "s", "t", "u", "v", "w", "x", "y", "z"]

upper_case = ["A", "B", "C", "D", "E", "F", "G", "H" "I", "J", "K", "L", "M", "N", "O", "P", "Q", "R", "S", "T", "U", "V", "W", "X", "Y", "Z"]

punctuation = [".", ",", "!", "?", "'", '"', "(", ")", " ", "-", ";", ":"]

2. Make one long list containing all the symbols.  Find the length of this list.


In [62]:
all_symbols = lower_case + upper_case + punctuation
type(all_symbols)
len(all_symbols)

63

3. Repeat the previous steps, but instead of using *list* type objects, use **numpy arrays**.  Don't forget to **import** the numpy library!


In [70]:
import numpy as np
lower_case = np.array(["a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "k", "l", "m", "n", "o", "p", "q", "r", "s", "t", "u", "v", "w", "x", "y", "z"])

upper_case = np.array(["A", "B", "C", "D", "E", "F", "G", "H", "I", "J", "K", "L", "M", "N", "O", "P", "Q", "R", "S", "T", "U", "V", "W", "X", "Y", "Z"])

punctuation = np.array([".", ",", "!", "?", "'", "\"", "(", ")", " ", "-", ";", ":"])

## Use np.concatenate to combine into long array
all_symbols = np.concatenate([lower_case, upper_case, punctuation])
all_symbols

array(['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's',
       't', 'u', 'v', 'w', 'x', 'y', 'z', 'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L',
       'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z', '.', ',', '!', '?', "'",
       '"', '(', ')', ' ', '-', ';', ':'], dtype='<U1')

4. Make a **pandas data frame**, with one column called "Symbol" containing the `all_symbols` array.


In [78]:
import pandas as pd

symbols_df = pd.DataFrame(all_symbols, columns = ["Symbol"])

# symbols_df.rename(columns = {0:"Symbol"})
symbols_df

Symbol
0       a
1       b
2       c
3       d
4       e
..    ...
59      )
60       
61      -
62      ;
63      :

[64 rows x 1 columns]

## Part Two: Decoding the secret message.


In [79]:
top_secret = pd.read_csv("https://www.dropbox.com/s/k72h1zewk4gtqep/PA_Secret_Code?dl=1", header=None)

top_secret = top_secret['0'].to_numpy()

KeyError: '0'

### Hint

Recall that basic math operations are **vectorized**.  This means you don't have to repeat the process for each item in a list; you can apply the math to the whole thing at once.

### Instructions

By altering this top secret set of numbers, you will be able to create a message.  Write your own code to complete the steps below.

1. Add 14 to every number.
2. Multiply every number by 18, then subtract 257.
3. Exponentiate every number.  (That is, do e^[number].  You may have to Google how to do this!)
4. Square every number.

**Checkpoint:** Headquarters has informed you that at this stage of decoding, there should be 352 numbers in the secret message that are below 17.  Write code to check if this is true for your data.


5. Turn your one-dimensional array of numbers into a two-dimensional array (matrix) with 5 columns.
6. Separately from your top secret numbers, create an array of all the even numbers between 1 and 382. Name it "evens". That is, "evens" should contain 2, 4, 6, 8 ..., 382.
7. Subtract the "evens" vector from the first column of your secret message matrix.
8. Subtract 100 from all numbers in the 18-24th rows of the 3rd column.
9. Multiply all numbers in the 4th and 5th column by 2.
10. Turn your matrix back into a one dimensional array.


**Checkpoint:** Headquarters has informed you that at this stage of decoding, all numbers in indices 500 and beyond are below 100.  Write code to check if your numbers fulfill this requirement.

11. Take the square root of all numbers in indices 37 to 464.
12. Round all numbers to the nearest whole number.
13. Replace all instances of the number 39 with 20.

**Checkpoint:** Headquarters has informed you that your final message should have 344 even numbers.  Write code to check if your numbers fulfill this requirement.

## Part 3: The secret message!

Run the following to discover the final message!  Google the first line of this message, if you do not recognize it, to see what it is.


In [ ]:
''.join(map(str, symbols_df.loc[top_secret]['Symbol']))